### PR TITLE
Added customize world button and started refactoring things

### PR DESCRIPTION
**Pull request by topisani**
we have world specific configs now.

all special structure classes are disabled

No ores or structures are currently generating

Basically the idea is to replace a lot of the code we have copied from vanilla with 1.10 code, instead of the 1.7 code that it has been made up off so far. That has removed a lot of config options and such, but i htink thats for the better

This is very much a WIP and partly here for discussion. Im hoping some of it might help on performance too

**PS:** the commit message is a mistake

### CODE REVIEW

File: src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java [hunk 1]
```diff
@@ -0,0 +1,815 @@
...
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+/**
+ * @author topisani
+ */
+@SideOnly(Side.CLIENT)
+public class GuiCustomizeWorldScreenRTG extends GuiScreen implements GuiSlider.FormatHelper, GuiPageButtonList.GuiResponder {
+    protected String title = "Customize World Settings";
+    protected String subtitle = "Page 1 of 3";
+    protected String pageTitle = "Basic Settings";
```
Comment (srs-bsns) [inline src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java:35]: All strings in this class will eventually all need to be localised and the .lang files updated.

File: src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+            case 156:
+            case 157:
+            case 158:
+            case 159:
+            case 160:
+            case 161:
+            default:
+                return String.format("%d", Integer.valueOf((int) p_175330_2_));
+            case 162:
+
```
Comment (srs-bsns) [inline src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java:150]: All of those empty cases can be removed.

File: src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+        this.list = new GuiPageButtonList(this.mc, this.width, this.height, 32, this.height - 32, 25, this, aguipagebuttonlist$guilistentry, aguipagebuttonlist$guilistentry1, aguipagebuttonlist$guilistentry2, aguipagebuttonlist$guilistentry3);
```
Comment (srs-bsns) [inline src/main/java/teamrtg/rtg/client/GuiCustomizeWorldScreenRTG.java:919]: These lines are excessively long and should be chopped down to an acceptable width limit. (In the old days that would mean a max of 72 columns, but that's a bit extreme for modern terminals.)

File: src/main/java/teamrtg/rtg/core/world/ChunkProviderRTG.java [hunk 4]
```diff
@@ -748,7 +605,7 @@ private void doPopulate(int x, int z) {
                     RealisticBiomeGenerator.forBiome(realisticBiome.getBiome()).rDecorate(rtgWorld, rand, worldX, worldZ, borderNoise[bn], river, hasPlacedVillageBlocks);
                 } else {
                     try {
-                        realisticBiome.getBiome().decorate(this.world, rand, worldCoords);
+                        realisticBiome.getBiome().decorate(this.world, rand, blockpos);
                     } catch (Exception e) {
                         RealisticBiomeGenerator.forBiome(realisticBiome.getBiome()).rDecorate(rtgWorld, rand, worldX, worldZ, borderNoise[bn], river, hasPlacedVillageBlocks);
```
Comment (srs-bsns) [inline src/main/java/teamrtg/rtg/core/world/ChunkProviderRTG.java:610]: There are plently of long lines in this class that can be chopped down for better readability.

File: src/main/java/teamrtg/rtg/core/world/ChunkProviderRTG.java [hunk 5]
```diff
@@ -160,24 +149,21 @@ public ChunkProviderRTG(World worldIn, long l) {
...
+            villageGenerator = (MapGenVillage) net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(villageGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.VILLAGE);
+            mineshaftGenerator = (MapGenMineshaft) net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(mineshaftGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.MINESHAFT);
+            scatteredFeatureGenerator = (MapGenScatteredFeature) net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(scatteredFeatureGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.SCATTERED_FEATURE);
+            ravineGenerator = net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(ravineGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.RAVINE);
+            oceanMonumentGenerator = (StructureOceanMonument) net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(oceanMonumentGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.OCEAN_MONUMENT);
         }
```
Comment (srs-bsns) [inline src/main/java/teamrtg/rtg/core/world/ChunkProviderRTG.java:160]: More long lines that can be chopped.

**Comment by srs-bsns** [comment]
The theme for this PR :smirk:

long_lines

**Comment by topisani** [comment]
That GUI class is directly copied from vanilla. The original intention was not to do world specific configs, much less a GUI screen, but to take another swing at performance. That's just happened to bring along the world specific stuff, and I thought why not quickly bring along the GUI from vanilla then. That's why that class looks like shit

**Comment by srs-bsns** [comment]
Heh. Those line comments weren't to criticise or anything, just a future reference for clean-up.

File: build.properties [hunk 6]
```diff
@@ -9,10 +10,10 @@ mod_url=https://github.com/Team-RTG/Realistic-Terrain-Generation
 mod_author="Team RTG"
 mod_creds=Forked from RWG by ted80. OpenSimplex Noise by KdotJPG.
 mod_logo=assets/rtg/logo.png
-mcf_version=12.17.0.1976
+mcf_version=12.18.0.2010
```
Comment (srs-bsns) [inline build.properties]: `12.18.1.2011` is current recommended.
`mc_version` should be updated to `1.10.2`

**Comment by whichonespink44** [comment]
@srs-bsns Feel free to keep adding line notes this and other PRs, and I would encourage others to do the same in the future. Code review's always a good thing, and the changes don't necessarily need to be made before the PR is merged, but at least the notes are there for future reference.

As for this PR itself, like I said in Discord... I've been a fan of config GUI screens ever since I saw Plant Mega Pack's robust system, so I'm all for this. And @topisani I get the reason for bringing it across now whilst you were in there, so that's all cool too. Maybe the config stuff could be something that @srs-bsns or @VelocityRa pick up when it's ready since it's only really you and @Zeno410 who can do the more advanced world-genny stuff at the moment (btw, @VelocityRa, feel free to add yourself to that list if you think you're up for it... and maybe I'll be able to add myself to that list eventually as well, but I need to get a little more familiar with the new code, which I should be able to do now that I've gone through the 1.7.10 -> 1.9.4 porting process).

As for the general approach of 'starting from scratch' with some of the vanilla systems, I think that's a good approach. There's a lot of legacy functionality from 1.7.10 that doesn't belong in v2 (scattered features & villages for example), so I'm all for trimming the fat so to speak, especially when we're trying to improve performance (every little bit helps, etc).

TL;DR
sw33333333333t
:stuck_out_tongue:

**Comment by topisani** [comment]
oh yeah, i just took a look at that code and wanted to make sure you knew that was the swedes and not me ;) The line comments are a great idea, and i agree, i probably shouldnt be using my little time to do the config gui. So if anyone is up for it, go ahead, its probably easier than it seems!

**Comment by whichonespink44** [comment]
@topisani What's the latest with this PR? I'm thinking of merging the BOP- and AC-support branches tonight. Would that interfere with what you're doing?

**Comment by topisani** [comment]
structures and ores are not generating, but feel free to merge those other ones

**Comment by topisani** [comment]
It is?

**Comment by VelocityRa** [comment]
TEST

**Comment by topisani** [comment]
are you there?

**Comment by VelocityRa** [comment]
Like I've expressed in Discord (maybe you missed it?), I think that it's time for this to be merged.

**Comment by whichonespink44** [comment]
Cool, should I try to resolve the conflicts in a separate branch, or is that something you guys are gonna do?

**Comment by VelocityRa** [comment]
I'm pretty sure this https://github.com/Team-RTG/Realistic-Terrain-Generation/pull/845/commits/84da9be4e21054ff3539eaf6c012668b08eeb47f broke it, so it's probably better if @topisani did the merge.

**Comment by whichonespink44** [comment]
@topisani @VelocityRa I'm gonna close this, but feel free to PR to the `RTG2-dev` branch when it's ready.
